rustsec/src/repository/git/modification_time.rs

```diff
@@ -0,0 +1,56 @@
+use crate::error::Error;
+use git2::Time;
+use std::{cmp::max, collections::HashMap, path::PathBuf};
+
+use super::GitPath;
+
+/// Tracks the time of latest modification of files in git.
+pub struct GitModificationTimes {
+    mtimes: HashMap<PathBuf, Time>,
+}
+
+impl GitModificationTimes {
+    /// Performance: collects all modification times on creation
+    /// and caches them. This is more efficient for looking up lots of files,
+    /// but wasteful if you just need to look up a couple files.
+    pub fn new(repo: &super::Repository) -> Result<Self, Error> {
+        // Sadly I had to hand-roll this; there is no good off-the-shelf impl.
+        // libgit2 has had a feature request for this for over a decade:
+        // https://github.com/libgit2/libgit2/issues/495
+        // as does git2-rs: https://github.com/rust-lang/git2-rs/issues/588
+        // To make sure this works I've verified it against a naive shell script using `git log`
+        // as well as `git whatchanged`
+        let mut mtimes: HashMap<PathBuf, Time> = HashMap::new();
+        let repo = git2::Repository::open(repo.path())?;
+        let mut revwalk = repo.revwalk()?;
+        revwalk.set_sorting(git2::Sort::TIME)?;
+        revwalk.push_head()?;
+        for commit_id in revwalk {
+            let commit_id = commit_id?;
+            let commit = repo.find_commit(commit_id)?;
+            // Ignore merge commits (2+ parents) because that's what 'git whatchanged' does.
+            // Ignore commit with 0 parents (initial commit) because there's nothing to diff against
```

Comment:
Would that also mean that if a file is touched only in the initial commit it would be missing entirely? Could that be relevant to the generated advisory output in the form of a missing advisory?
Asked differently, would it hurt to diff against an empty tree (`None`) in that case to also fetch these paths just in case?

Reply:
In the general case you're probably right, but this is not relevant for the advisory database which was initialized with a README, and which has been touched since.

Reply:
I've opened https://github.com/rustsec/rustsec/pull/440 to address this. Thanks again!